sympy/core/numbers.py

```diff
@@ -1745,93 +1759,95 @@ def __eq__(self, other):
             other = _sympify(other)
         except SympifyError:
             return NotImplemented
+        if other.is_zero:
+            return self.is_zero
+        if self.is_zero:
+            return other.is_zero
         if other.is_NumberSymbol:
             if other.is_irrational:
                 return False
             return other.__eq__(self)
-        if other.is_Number:
-            if other.is_Rational:
-                # a Rational is always in reduced form so will never be 2/4
-                # so we can just check equivalence of args
-                return self.p == other.p and self.q == other.q
-            if other.is_Float:
-                return mlib.mpf_eq(self._as_mpf_val(other._prec), other._mpf_)
+        if other.is_Rational:
+            # a Rational is always in reduced form so will never be 2/4
+            # so we can just check equivalence of args
+            return self.p == other.p and self.q == other.q
+        if other.is_Float:
+            s, m, t = other._mpf_[:3]
+            if s:
+                m = -m
+            if not t:
+                # other is an odd integer
+                if not self.is_Integer or self.is_even:
+                    return False
+                return m == self.p
+            if t > 0:
+                # other is an even integer
+                if not self.is_Integer:
+                    return False
+                # does m*2**t == self.p
+                _2t, r = divmod(self.p, m)
+                return not r and _2t >> t == 1
+            # does non-integer s*m/2**-t = p/q?
+            if self.is_Integer:
+                return False
+            return m == self.p and self.q >> -t == 1
         return False
 
     def __ne__(self, other):
         return not self == other
 
-    def __gt__(self, other):
+    def _Rrel(self, other, attr):
+        # if you want self < other, pass self, other, __gt__
         try:
             other = _sympify(other)
         except SympifyError:
             raise TypeError("Invalid comparison %s > %s" % (self, other))
-        if other.is_NumberSymbol:
-            return other.__lt__(self)
-        expr = self
         if other.is_Number:
-            if other.is_Rational:
-                return _sympify(bool(self.p*other.q > self.q*other.p))
-            if other.is_Float:
-                return _sympify(bool(mlib.mpf_gt(
-                    self._as_mpf_val(other._prec), other._mpf_)))
-        elif other.is_number and other.is_real:
-            expr, other = Integer(self.p), self.q*other
-        return Expr.__gt__(expr, other)
+            op = None
+            if other.is_NumberSymbol:
+                op = getattr(other, attr)
+            elif other.is_Float:
+                op = getattr(other, attr)
+            elif other.is_Rational:
+                self, other = self.p*other.q, self.q*other.p
+                other = _sympify(other)
```

Comment:
Should both `self` and `other` be made Integers?

Reply:
`self` is already the SymPy class type so this would be redundant.